src/iter.rs

```diff
@@ -20,62 +21,80 @@ pub struct MiniconfIter<M, S, const L: usize, P> {
     ///
     /// It may be None to indicate unknown length.
     count: Option<usize>,
-}
 
-impl<M, S, const L: usize, P> Default for MiniconfIter<M, S, L, P> {
-    fn default() -> Self {
-        Self {
-            count: None,
-            marker: PhantomData,
-            state: [0; L],
-        }
-    }
+    /// The separator before each name.
+    separator: &'a str,
 }
 
-impl<M: SerDe<S>, S, const L: usize, P> MiniconfIter<M, S, L, P> {
-    pub fn metadata() -> Result<Metadata, IterError> {
-        let meta = M::metadata(M::SEPARATOR.len_utf8());
+impl<'a, M: Miniconf + ?Sized, const L: usize, P> PathIter<'a, M, L, P> {
+    pub(crate) fn new(separator: &'a str) -> core::result::Result<Self, Error<SliceShort>> {
+        let meta = M::metadata();
         if L < meta.max_depth {
-            return Err(IterError::Depth);
+            return Err(Error::Inner(SliceShort));
         }
-        Ok(meta)
+        let mut s = Self::new_unchecked(separator);
+        s.count = Some(meta.count);
+        Ok(s)
     }
 
-    pub fn new() -> Result<Self, IterError> {
-        let meta = Self::metadata()?;
-        Ok(Self {
-            count: Some(meta.count),
-            ..Default::default()
-        })
+    pub(crate) fn new_unchecked(separator: &'a str) -> Self {
+        Self {
+            count: None,
+            separator,
+            state: [0; L],
+            m: PhantomData,
+            p: PhantomData,
+        }
     }
 }
 
-impl<M: SerDe<S>, S, const L: usize, P: Write + Default> Iterator for MiniconfIter<M, S, L, P> {
+impl<'a, M, const L: usize, P> Iterator for PathIter<'a, M, L, P>
+where
+    M: Miniconf + ?Sized,
+    P: Write + Default,
+{
     type Item = P;
 
     fn next(&mut self) -> Option<Self::Item> {
         let mut path = Self::Item::default();
 
         loop {
```

Comment:
Why the loop if we're returning the very first M::path() value?

Reply:
We are not. There is exactly one `continue` branch.

Reply:
Ah yes, I see now, thanks!